akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala

```diff
@@ -1,129 +1,101 @@
-            case _ =>
-            // ok, might happen after standby recovery, worker state is not persisted
-          }
-
-        val listingResponseAdapter = ctx.messageAdapter[Receptionist.Listing](UpdatedWorkers)
-        ctx.system.receptionist ! Receptionist.Subscribe(WorkManager.WorkerServiceKey, listingResponseAdapter)
+        }
 
         EventSourcedBehavior[Command, WorkDomainEvent, WorkState](
           persistenceId = PersistenceId.ofUniqueId("master"),
           emptyState = WorkState.empty,
           commandHandler = (workState, command) => {
             command match {
-              case UpdatedWorkers(listing) =>
-                val newWorkerList: Set[ActorRef[WorkerCommand]] =
-                  listing.allServiceInstances(WorkManager.WorkerServiceKey)
-
-                var events = ListBuffer.empty[WorkDomainEvent]
-                val removedWorkers = workers.keySet.diff(newWorkerList)
-                val newWorkers = newWorkerList.diff(workers.keySet)
-                // for each removed, check if they had work in progress
-                removedWorkers.foreach { removedWorkers =>
-                  workers(removedWorkers) match {
-                    case WorkerState(_, Idle) =>
-                    // that's fine, nothing in progress
-                    case WorkerState(_, Busy(workId, _)) =>
-                      // work considered failed
-                      events += WorkerFailed(workId)
-                  }
-                  workers -= removedWorkers
-                }
-
-                // update the current members
-                newWorkers.foreach { newWorker =>
-                  workers += (newWorker -> WorkerState(newWorker, Idle))
-                }
-                ctx.log.info("Workers updated. Removed workers: {}. New workers: {}", removedWorkers, newWorkers)
-                ctx.log.info("All workers: {}", workers)
-
-                if (workState.hasWork) {
-                  newWorkers.foreach { newWorker =>
-                    newWorker ! WorkAvailable
-                  }
-                }
-                if (events.nonEmpty) {
-                  Effect.persist(events.toList).thenRun(notifyWorkers)
-                } else {
-                  Effect.none
+              case RequestNextWrapper(rn) =>
+                ctx.log.info("work request: {}. Current: {}", rn, requestNext.size)
+                if (requestNext.isDefined) {
+                  throw new IllegalStateException(s"Request next when there is already demand ${rn}, ${requestNext}")
                 }
+                requestNext = Some(rn)
+                tryStartWork(workState)
+              case TryStartWork() =>
+                tryStartWork(workState)
+              case WorkIsDone(workId) =>
+                 Effect.persist[WorkDomainEvent, WorkState](WorkCompleted(workId)).thenRun { newState =>
+                   // FIXME, how to get the result back?
+                   // publish it from the worker?
```

Comment:
WorkResultConsumer should be updated to new pub/sub, if we keep that.
However, that communication isn't reliable so kind of defeats the purpose of the sample if that is the only way to retrieve the result. I think a real system would store the result (from the workers) so that clients can ask for it (several times if needed). That was my approach in the doc sample https://doc.akka.io/docs/akka/current/typed/reliable-delivery.html#work-pulling
Replying back with a workId. Here I think that reply can be immediately when the WorkAccepted has been persisted, which is exactly what it does already.

Reply:
Maybe we can just leave out the aspect of how to retrieve the work result from the sample, to keep it simple.